### PR TITLE
Revise race pair

### PR DESCRIPTION
**Pull request by RaasAhsan**
- [x] Fix laws
- [x] Fix `Gen` instances

### CODE REVIEW

File: core/shared/src/main/scala/cats/effect/IO.scala [hunk 1]
```diff
@@ -46,8 +46,24 @@ sealed abstract class IO[+A] private () extends IOPlatform[A] {
 
   def both[B](that: IO[B]): IO[(A, B)] =
```
Comment (RaasAhsan) [inline core/shared/src/main/scala/cats/effect/IO.scala:53]: Seems like we could have a `bothOutcome` in a similar fashion to `raceOutcome`

File: core/shared/src/main/scala/cats/effect/IO.scala [hunk 2]
```diff
@@ -101,13 +117,43 @@ sealed abstract class IO[+A] private () extends IOPlatform[A] {
...
+        case Outcome.Completed(fa) => f.cancel *> fa.map(Left.apply)
+        case Outcome.Errored(ea) => f.join.flatMap {
+          case Outcome.Completed(fb) => fb.map(Right.apply)
+          case Outcome.Errored(eb) => IO.raiseError(eb)
+          case Outcome.Canceled() => IO.raiseError(ea)
+        }
+        case Outcome.Canceled() => f.join.flatMap {
+          case Outcome.Completed(fb) => fb.map(Right.apply)
+          case Outcome.Errored(eb) => IO.raiseError(eb)
+          case Outcome.Canceled() => IO.canceled *> IO.never
```
Comment (RaasAhsan) [inline core/shared/src/main/scala/cats/effect/IO.scala]: `IO.canceled *> IO.never` can deadlock if we're masked here

File: core/shared/src/main/scala/cats/effect/IO.scala [hunk 3]
```diff
@@ -101,13 +121,50 @@ sealed abstract class IO[+A] private () extends IOPlatform[A] {
 
   def race[B](that: IO[B]): IO[Either[A, B]] =
     racePair(that).flatMap {
-      case Left((a, f)) => f.cancel.as(Left(a))
-      case Right((f, b)) => f.cancel.as(Right(b))
+      case Left((oc, f)) =>
+        oc match {
+          case Outcome.Completed(fa) => f.cancel *> fa.map(Left.apply)
```
Comment (djspiewak) [inline core/shared/src/main/scala/cats/effect/IO.scala]:
```suggestion
          case Outcome.Completed(fa) => f.cancel *> fa.map(Left(_))
```

And analogously for the others. I think it's just a bit cleaner.

File: core/shared/src/main/scala/cats/effect/IO.scala [hunk 4]
```diff
@@ -101,13 +121,50 @@ sealed abstract class IO[+A] private () extends IOPlatform[A] {
...
+              case Outcome.Canceled() => IO.canceled *> IO.never
+            }
+        }
+      case Right((f, oc)) =>
+        oc match {
+          case Outcome.Completed(fb) => f.cancel *> fb.map(Right.apply)
+          case Outcome.Errored(eb) =>
+            f.join.flatMap {
+              case Outcome.Completed(fa) => fa.map(Left.apply)
+              case Outcome.Errored(ea) => IO.raiseError(ea)
```
Comment (djspiewak) [inline core/shared/src/main/scala/cats/effect/IO.scala]: BTW this is a slightly different semantic than before, where the *first* error would be emitted. I actually think we're better off just failing the race if one errors. I haven't checked on what ce2's semantics are here, but it just seems odd to me (in retrospect) that `race(raiseError(e), never)` would be a deadlock.

File: core/shared/src/main/scala/cats/effect/IO.scala [hunk 5]
```diff
@@ -101,13 +121,50 @@ sealed abstract class IO[+A] private () extends IOPlatform[A] {
...
+              case Outcome.Canceled() => IO.canceled *> IO.never
+            }
+        }
+    }
+
+  def raceOutcome[B](that: IO[B])
+      : IO[Either[Outcome[IO, Throwable, A @uncheckedVariance], Outcome[IO, Throwable, B]]] =
```
Comment (djspiewak) [inline core/shared/src/main/scala/cats/effect/IO.scala]: Still not convinced this is actually useful. :-)

File: core/shared/src/main/scala/cats/effect/IOFiber.scala [hunk 6]
```diff
@@ -442,79 +438,30 @@ private[effect] final class IOFiber[A](name: String, scheduler: unsafe.Scheduler
...
-                        if (error != null) {
-                          // we were canceled, and the other errored, so use its error
-                          cb(Left(error))
-                        }
-                      }
-                    }
-                  }
+            val next = IO.async[Either[
+              (Outcome[IO, Throwable, Any], Fiber[IO, Throwable, Any]),
+              (Fiber[IO, Throwable, Any], Outcome[IO, Throwable, Any])]] { cb =>
```
Comment (djspiewak) [inline core/shared/src/main/scala/cats/effect/IOFiber.scala]: We should like, make a type alias for this.

File: core/shared/src/main/scala/cats/effect/IOFiber.scala [hunk 7]
```diff
@@ -442,79 +438,30 @@ private[effect] final class IOFiber[A](name: String, scheduler: unsafe.Scheduler
...
+                val ec = currentCtx
 
-                  execute(ec)(() => fiberA.run(cur.ioa, ec, masks))
-                  execute(ec)(() => fiberB.run(cur.iob, ec, masks))
+                execute(ec)(() => fiberA.run(cur.ioa, ec))
+                execute(ec)(() => fiberB.run(cur.iob, ec))
 
-                  Some(fiberA.cancel *> fiberB.cancel)
-                }
+                Some(fiberA.cancel *> fiberB.cancel)
```
Comment (djspiewak) [inline core/shared/src/main/scala/cats/effect/IOFiber.scala]: Random thing that should probably be fixed… this should be run in parallel. We really want `IO.both(fiberA.cancel, fiberB.cancel).void`.

File: testkit/shared/src/main/scala/cats/effect/testkit/TimeT.scala [hunk 8]
```diff
@@ -40,6 +40,11 @@ object TimeT {
   def liftF[F[_], A](fa: F[A]): TimeT[F, A] =
     Kleisli.liftF(fa)
 
+  def liftK[F[_], A]: F ~> TimeT[F, *] =
```
Comment (djspiewak) [inline testkit/shared/src/main/scala/cats/effect/testkit/TimeT.scala]: `A`?

**Review by djspiewak — CHANGES_REQUESTED** [review]
My suggestion for the laws would be to remove most of the `race` laws and talk about `racePair` with a couple new ones instead. Also, to that end, `PureConc` needs to be updated.

File: testkit/shared/src/main/scala/cats/effect/testkit/TimeT.scala (hunk 8 above)
Comment (RaasAhsan) [reply testkit/shared/src/main/scala/cats/effect/testkit/TimeT.scala]: I copied the type signature from `Kleisli.liftK` and didn't realize it was used 😅

File: core/shared/src/main/scala/cats/effect/IOFiber.scala (hunk 6 above)
Comment (RaasAhsan) [reply core/shared/src/main/scala/cats/effect/IOFiber.scala]: I was thinking about this too, maybe a specialized `Outcome` and `Fiber` for `IO` makes sense. When we're talking about either of those in the context of `IO`, `Throwable` is pretty much implied.

`IOOutcome`, `IOFiber` (this ones already taken but we can come up with better names for the implementation I think)

File: core/shared/src/main/scala/cats/effect/IO.scala (hunk 4 above)
Comment (RaasAhsan) [reply core/shared/src/main/scala/cats/effect/IO.scala]: I tried to mimic the current CE3 semantics, it seems like the second error is reported https://github.com/typelevel/cats-effect/blob/series/3.x/core/shared/src/main/scala/cats/effect/IOFiber.scala#L484

File: core/shared/src/main/scala/cats/effect/IO.scala (hunk 5 above)
Comment (RaasAhsan) [reply core/shared/src/main/scala/cats/effect/IO.scala]: I'm sure somebody could come up with a contrived example... but what do you think about `bothOutcome`?

File: core/shared/src/main/scala/cats/effect/IO.scala (hunk 4 above)
Comment (djspiewak) [reply core/shared/src/main/scala/cats/effect/IO.scala]: Oh interesting! That wasn't intended. I think. I'd have to look at `PureConc`. In any case though, I think it's better to make the error report immediate; in other words, I think the old ce3 semantics are wrong.

File: core/shared/src/main/scala/cats/effect/IO.scala (hunk 5 above)
Comment (djspiewak) [reply core/shared/src/main/scala/cats/effect/IO.scala]: `bothOutcome` feels a bit more useful. :-) If nothing else, `both(raiseError(e1), raiseError(e2))` is inherently weird, whereas it works fine with `bothOutcome`

**Comment by RaasAhsan** [comment]
@djspiewak Updated the `PureConc.racePair` implementation, let me know if it looks right :)

The law for `racePair` aren't very elegant to write because of `Outcome` :) . My first attempt was something like this, but it doesn't pass exhaustively.
```scala
def racePairNeverIdentityLeft[A](fa: F[A]) = {
    val identity = F.racePair(F.never[Unit], fa).flatMap {
      case Left(_) => F.never[Outcome[F, E, A]]
      case Right((_, oc)) => F.pure(oc)
    }

    fa.map(a => Outcome.completed[F, E, A](F.pure(a))) <-> identity
  }
```

**Comment by RaasAhsan** [comment]
Ah, well the concurrent laws for `PureConc` are failing which seems to suggest that the `racePair` implementation doesn't work.

```
[error] x concurrent.uncancelable race displaces canceled
[error]  Falsified after 0 passed tests.
[error]  > Labels of failing property:Expected: "Completed(None)\u000a│\u000a├ Notify 0x37D631D2\u000a├ Notify 0x37D631D2\u000a├ Notify 0x37D631D2\u000a╰ Done"
[error]  Received: "Completed(Some(()))\u000a│\u000a├ Notify 0x22CBF51E\u000a├ Notify 0x58164174\u000a├ Fork\u000a│ ╰╮\u000a│  ├ Notify 0x58164174\u000a│  ├ Notify 0x58164174\u000a│  ╰ Done\u000a│  \u000a├ Notify 0x73D34EA4\u000a├ Fork\u000a│ ╰╮\u000a│  ├ Notify 0x73D34EA4\u000a│  ├ Notify 0x73D34EA4\u000a│  ├ Notify 0x4E172C51\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a│  ├ Await 0x7C43D089\u000a\u000a...\u000a\u000a..."
[error]  The seed is P36o_rk8LlGX5TCkjlFJepb7OtUwwbENSNvQKwGZm_D=
[info] + concurrent.uncancelable race poll canceled identity (left)
[error] x concurrent.uncancelable race poll canceled identity (right)
[error]  Falsified after 0 passed tests.
[error]  > Labels of failing property:Expected: "Completed(None)\u000a│\u000a├ Notify 0xE638D54\u000a├ Notify 0xE638D54\u000a├ Notify 0xE638D54\u000a╰ Done"
[error]  Received: "Completed(Some(Left(1299570979)))\u000a│\u000a├ Notify 0x3AE912B2\u000a├ Notify 0x165C8DD4\u000a├ Fork\u000a│ ╰╮\u000a│  ├ Notify 0x165C8DD4\u000a│  ├ Notify 0x165C8DD4\u000a│  ├ Notify 0x4534522C\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a│  ├ Await 0x57AF7EF3\u000a\u000a...\u000a\u000a..."> ARG_0: "Completed(None)\u000a│\u000a├ Notify 0x4023AF0\u000a├ Notify 0x4023AF0\u000a├ Notify 0x4023AF0\u000a╰ Done"
[error]  The seed is ZhNu4goN-9O4XlJ2JWTzFrjI09c9Ih_XDQ0umO1IbEG=
```

I'm not really sure how to debug this thing without drowning a lot of time into figuring out how it really works, so let me know if you can spot what I looked over

**Comment by djspiewak** [comment]
> Ah, well the concurrent laws for PureConc are failing which seems to suggest that the racePair implementation doesn't work.

Some of the old laws need to be removed! :-) There are a bunch of laws which claim that `uncancelable` should distribute over `racePair` and such. All of those need to go.

**Comment by RaasAhsan** [comment]
@djspiewak I think all the `racePair` variants are going to have to wrapped in `uncancelable`, or else we might leak resources. Not sure if there are any reasonable regions we can poll inside. 

I want to get this branch merged as soon as fix these lingering `PureConc` issues, but I have it on my TODO to get to that afterwards. We should probably do a sweep of both `IO` and `Concurrent` machinery to see we're not missing any of these cases.

**Comment by RaasAhsan** [comment]
I managed to isolate this failing `Align` test case for `PureConc`. It seems like the issue comes from either `both` or `racePair`. I've already made all the substitutions, but here's a reduced case that illustrate where things seem to go wrong:

```scala
// The following two are NOT the same when they participate in a `racePair`
fa.align(fb.map(identity))
fa.align(fb)

// If you make the appropriate substitutions, you end up here
val a = (ParallelF.value(ParallelF(fa).map(identity)).attempt, ParallelF.value(ParallelF(fb).map(identity)).attempt)
            .parMapN((ea, eb) => catsStdInstancesForEither.alignWith(ea, eb)(identity))
            .flatMap(C.fromEither)
// `ParallelF.value(ParallelF(fb).map(identity))` is the term of interest. that map call eventually makes its way to `both` and `racePair`, so something is wrong there
```

Here are the renderings of the different effects:
```scala
// fa
Completed(Some(1))
│
├ Notify 0x2E408AB9
├ Cede
├ Notify 0x15A9C7E3
╰ Pure Completed(1)

// fb
Completed(None)
│
├ Notify 0x3641C43B
├ Notify 0x3641C43B
├ Notify 0x3641C43B
╰ Done

// ParallelF.value(ParallelF(fa).map(identity).align(ParallelF(fb)))
Completed(Some(Ior.Left(1)))
│
├ Notify 0x1C152686
├ Notify 0x13BD612B
├ Fork
│ ╰╮
│  ├ Notify 0x13BD612B
│  ├ Notify 0x13BD612B
│  ├ Notify 0x5612A9E1
│  ├ Fork
│  │ ╰╮
│  │  ├ Notify 0x5612A9E1
│  │  ├ Notify 0x5612A9E1
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53
│  │  ├ Await 0x951FB53

...

...

...

// ParallelF.value(ParallelF(fa).map(identity).align(ParallelF(fb).map(identity)))
Completed(None)
│
├ Notify 0x20A079E1
├ Notify 0xCEE2D38
├ Fork
│ ╰╮
│  ├ Notify 0xCEE2D38
│  ├ Notify 0xCEE2D38
│  ├ Notify 0x2E102352
│  ├ Fork
│  │ ╰╮
│  │  ├ Notify 0x2E102352
│  │  ├ Notify 0x2E102352
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5
│  │  ├ Await 0x229EFF5

...

...

...
```

The failing seed for the `alignHomomorphism` test is `-ETrHGyCTzib58PxGaFyumO3n6Of8RjswGuqSDhQvnN=`.

**Comment by djspiewak** [comment]
So, `fa` is something like `F.cede.as(1)`, while `fb` could be either `F.canceled` or `F.never`. Really nice minimization btw! I'll take a look at it in detail as soon as I can.

**Pull request merged**